File: admin/src/types/analytics.ts
// Types for the sales analytics dashboard (GET /analytics/dashboard).

export interface AnalyticsKpis {
  gross_sales: number;
  net_sales: number;
  transactions: number;
  items_sold: number;
  avg_ticket: number;
  gross_sales_delta: number | null;
  net_sales_delta: number | null;
  transactions_delta: number | null;
  items_sold_delta: number | null;
  avg_ticket_delta: number | null;
}

export interface PaymentBreakdownItem {
  method: string;
  amount: number;
}

export interface OutletSalesItem {
  outlet_id: string;
  outlet_name: string;
  net_sales: number;
  transactions: number;
}

export interface TrendPoint {
  date: string;
  net_sales: number;
  transactions: number;
}

export interface TopProductItem {
  product_id: string;
  name: string;
  quantity: number;
  revenue: number;
}

export interface DayOfWeekPoint {
  day: string;
  avg_sales: number;
  avg_transactions: number;
}

export interface HourlyPoint {
  hour: number;
  revenue: number;
  transactions: number;
  avg_sales: number;
}

export interface ConcentrationData {
  top3_pct: number;
  top5_pct: number;
  total_products: number;
}

export interface ScoopRatioData {
  single_qty: number;
  double_qty: number;
  single_pct: number;
  double_pct: number;
}

export interface AnalyticsDashboard {
  date_from: string;
  date_to: string;
  kpis: AnalyticsKpis;
  payments: PaymentBreakdownItem[];
  sales_by_outlet: OutletSalesItem[];
  trend: TrendPoint[];
  top_by_revenue: TopProductItem[];
  top_by_quantity: TopProductItem[];
  day_of_week: DayOfWeekPoint[];
  hourly: HourlyPoint[];
  concentration: ConcentrationData;
  scoop_ratio: ScoopRatioData;
}

export interface AnalyticsDashboardParams {
  outlet_id?: string;
  days?: number;
  from_date?: string;
  to_date?: string;
}

// Staff leaderboard, mapped from GET /reports/staff.
export interface StaffLeaderboardEntry {
  staff_id: string;
  name: string;
  transactions: number;
  revenue: number;
  average_ticket: number;
}

// ── Flavour analytics (GET /analytics/flavors, /analytics/flavor-rankings) ──

export type FlavorClassification = 'hero' | 'rising' | 'stable' | 'rotate_out';
export type RankClassification =
  | 'consistent_top'
  | 'rising'
  | 'stable'
  | 'dropping'
  | 'consistent_bottom';

export interface FlavorTrendPoint {
  bucket: string;
  qty: number;
}

export interface FlavorOutletQty {
  outlet_id: string;
  outlet_name: string;
  qty: number;
  share_pct: number;
}

export interface FlavorItem {
  name: string;
  total_qty: number;
  share_pct: number;
  rank: number;
  trend_pct: number;
  classification: FlavorClassification;
  days_sold: number;
  active_days: number;
  consistency_pct: number;
  trend: FlavorTrendPoint[];
  by_outlet: FlavorOutletQty[];
}

export interface ParetoPoint {
  name: string;
  quantity: number;
  share_pct: number;
  cumulative_pct: number;
}

export interface FlavorDowRow {
  name: string;
  quantities: number[]; // Monday..Sunday
}

export interface FlavorAnalysis {
  date_from: string;
  date_to: string;
  active_days: number;
  flavors: FlavorItem[];
  addons: FlavorItem[];
  pareto: ParetoPoint[];
  day_of_week: FlavorDowRow[];
}

export interface FlavorAnalysisParams extends AnalyticsDashboardParams {
  granularity?: 'daily' | 'weekly' | 'monthly';
}

export interface FlavorRankingItem {
  name: string;
  monthly_rank: (number | null)[];
  monthly_qty: number[];
  rank_change: number | null;
  avg_rank: number;
  classification: RankClassification;
}

export interface FlavorRankings {
  months: string[];
  flavors: FlavorRankingItem[];
}

File: backend/app/schemas/analytics.py
"""Response schemas for the sales analytics dashboard."""

from pydantic import BaseModel


class MonthEndOutletForecast(BaseModel):
    """One visible outlet's current earned vs projected full-month target."""

    outlet_id: str
    outlet_name: str
    projected_total: float
    earned_so_far: float
    remaining: float
    method: str = "run_rate_fallback"
    history_days: int = 0
    # Actual vs model-expected sales on completed days so far this month.
    # Positive => ahead of historical trend; negative => behind; None => no basis.
    pace_variance_pct: float | None = None


class AnalyticsKpis(BaseModel):
    """Headline totals for the selected period, with % change vs the previous period.

    Deltas are None when there is no comparison period (all-time view) or the
    previous period had no sales.
    """

    gross_sales: float
    net_sales: float
    transactions: int
    items_sold: int
    avg_ticket: float
    gross_sales_delta: float | None = None
    net_sales_delta: float | None = None
    transactions_delta: float | None = None
    items_sold_delta: float | None = None
    avg_ticket_delta: float | None = None
    # Month-end prediction (current month only, SGT).
    # projected_total: total net sales projected for the full month
    # earned_so_far:   actual net sales for the current month to date.
    # remaining:       projected_total - earned_so_far
    month_end_projected_total: float = 0.0
    month_end_earned_so_far: float = 0.0
    month_end_remaining: float = 0.0
    month_end_method: str = "run_rate_fallback"
    month_end_history_days: int = 0
    month_end_sample_count: int = 0
    # Per-outlet current/target breakdown for the current month (empty otherwise).
    month_end_outlets: list[MonthEndOutletForecast] = []
    # Group actual-vs-expected pace variance % for completed days this month.
    month_end_pace_variance_pct: float | None = None
    # Operational attach metrics over the selected range (waffle/drink attach
    # rate as % of paid orders; pints = units sold).
    waffle_attach_rate: float = 0.0
    drink_attach_rate: float = 0.0
    pints_sold: int = 0


class PaymentBreakdownItem(BaseModel):
    method: str
    amount: float


class OutletSalesItem(BaseModel):
    outlet_id: str
    outlet_name: str
    net_sales: float
    transactions: int


class TrendPoint(BaseModel):
    date: str
    net_sales: float
    transactions: int


class TopProductItem(BaseModel):
    product_id: str
    name: str
    quantity: int
    revenue: float


class DayOfWeekPoint(BaseModel):
    day: str
    avg_sales: float
    avg_transactions: float


class HourlyPoint(BaseModel):
    hour: int
    revenue: float
    transactions: int
    avg_sales: float


class ConcentrationData(BaseModel):
    top3_pct: float
    top5_pct: float
    total_products: int


class ScoopRatioData(BaseModel):
    single_qty: int
    double_qty: int
    single_pct: float
    double_pct: float


class RedemptionTypeBreakdown(BaseModel):
    type: str
    count: int
    value: float


class CdcRedemptions(BaseModel):
    """Government CDC amounts keyed in at checkout (orders.cdc_amount > 0)."""

    orders: int
    value: float


class VoucherRedemptions(BaseModel):
    """Scanned voucher redemptions from order_vouchers."""

    count: int
    value: float
    by_type: list[RedemptionTypeBreakdown]


class RedemptionsData(BaseModel):
    cdc: CdcRedemptions
    vouchers: VoucherRedemptions


class AnalyticsDashboardResponse(BaseModel):
    date_from: str
    date_to: str
    kpis: AnalyticsKpis
    payments: list[PaymentBreakdownItem]
    sales_by_outlet: list[OutletSalesItem]
    trend: list[TrendPoint]
    top_by_revenue: list[TopProductItem]
    top_by_quantity: list[TopProductItem]
    day_of_week: list[DayOfWeekPoint]
    hourly: list[HourlyPoint]
    concentration: ConcentrationData
    scoop_ratio: ScoopRatioData
    redemptions: RedemptionsData
